File: src/main/java/org/ankol/server/api/ClashApi.java
package org.ankol.server.api;

import cn.hutool.core.map.MapUtil;
import cn.hutool.http.HttpException;
import cn.hutool.http.HttpResponse;
import com.google.gson.reflect.TypeToken;
import org.ankol.server.api.entity.*;
import org.ankol.server.tools.GsonUtil;

public class ClashApi extends AbstractApi {
    public ClashApi(String authorization) {
        super(authorization);
    }

    public final ClanApi clan = new ClanApi(authorization);
    public final Player player = new Player(authorization);

    public static class ClanApi extends AbstractApi {
        public ClanApi(String authorization) {
            super(authorization);
        }

        /**
         * 通过部落标签获取单个氏族的信息，部落标签可以通过部落搜索找到。
         * <p>请注意，部落标签以哈希字符'#'开头，需要正确地进行URL编码才能在URL中发挥作用。</p>
         * <p>例如，部落标签'#2ABC'在URL中会变成'%232ABC'。</p>
         *
         * @param clanTag 部落标签
         * @return 单个部落信息
         */
        public Clan clan(String clanTag) {
            HttpResponse response = super.doGet(BASE_URL + "/clans/" + clanTag);
            if (response.isOk()) {
                return GsonUtil.fromJson(response.body(), Clan.class);
            }
            throw new HttpException("Error response status [{}] Message: {}", response.getStatus(), response.body());
        }

        /**
         * 获取部落当前的部落战信息
         *
         * @param clanTag 部落标签
         */
        public void currentWar(String clanTag) {

        }

        /**
         * 获取部落突袭周末数据
         *
         * @param clanTag 部落标签.
         * @param limit   响应中的返回项目限制数量.
         * @param after   Return only items that occur after this marker.
         *                Before marker can be found from the response, inside the 'paging' property.
         *                Note that only after or before can be specified for a request, not both.
         * @param before  Return only items that occur before this marker.
         *                Before marker can be found from the response, inside the 'paging' property.
         *                Note that only after or before can be specified for a request, not both.
         * @return {@link ItemResult <RaidSeason>}
         */
        public ItemResult<RaidSeason> capitalRaidSeasons(String clanTag, Integer limit, String after, String before) throws HttpException {
            HttpResponse response = super.doGet(BASE_URL + "/clans/" + clanTag + "/capitalraidseasons",
                    MapUtil.<String, Object>builder("limit", limit)
                            .put("after", after)
                            .put("before", before)
                            .build()
            );
            if (response.isOk()) {
                return GsonUtil.fromJson(response.body(), new TypeToken<>() {
                });
            }
            throw new HttpException("Error response status [{}] Message: {}", response.getStatus(), response.body());
        }

        /**
         * 获取部落突袭周末数据
         *
         * @param clanTag 部落标签.
         * @return {@link ItemResult < RaidSeason >}
         */
        public ItemResult<RaidSeason> capitalRaidSeasons(String clanTag) throws HttpException {
            return this.capitalRaidSeasons(clanTag, null, null, null);
        }

        /**
         * 获取部落突袭周末数据
         *
         * @param clanTag 部落标签.
         * @param limit   响应中的返回项目限制数量.
         * @return {@link ItemResult < RaidSeason >}
         */
        public ItemResult<RaidSeason> capitalRaidSeasons(String clanTag, Integer limit) throws HttpException {
            return this.capitalRaidSeasons(clanTag, limit, null, null);
        }

        /**
         * 列举部落成员
         *
         * @param clanTag 部落标签
         * @param limit   响应中的返回项目限制数量.
         * @param after   Return only items that occur after this marker.
         *                Before marker can be found from the response, inside the 'paging' property.
         *                Note that only after or before can be specified for a request, not both.
         * @param before  Return only items that occur before this marker.
         *                Before marker can be found from the response, inside the 'paging' property.
         *                Note that only after or before can be specified for a request, not both.
         * @return {@link ItemResult <  ClanMember  >}
         */
        public ItemResult<ClanMember> listMembers(String clanTag, Integer limit, String after, String before) {
            try (HttpResponse response = super.doGet(BASE_URL + "/clans/" + clanTag + "/members",
                    MapUtil.<String, Object>builder("limit", limit)
                            .put("after", after)
                            .put("before", before)
                            .build()
            )) {
                if (response.isOk()) {
                    return GsonUtil.fromJson(response.body(), new TypeToken<>() {
                    });
                } else {
                    throw new HttpException(response.body());
                }
            }
        }

        /**
         * 列举部落成员
         *
         * @param clanTag 部落标签
         * @return {@link ItemResult < ClanMember >}
         */
        public ItemResult<ClanMember> listMembers(String clanTag) {
            return this.listMembers(clanTag, null, null, null);
        }
    }

    public static class Player extends AbstractApi {
        public Player(String authorization) {
            super(authorization);
        }

        /**
         * 通过玩家标签获取单个玩家的信息。
         * 玩家标签可以在游戏中找到，也可以从氏族成员名单中找到。
         * 注意，玩家标签以哈希字符'#'开头，需要正确地进行URL编码才能在URL中发挥作用。
         * 例如，玩家标签'#2ABC'在URL中会变成'%232ABC'。
         *
         * @param playerTag 玩家标签
         * @return {@link ClanPlayer}
         */
        public ClanPlayer getPlayerDetail(String playerTag) {
            try (HttpResponse response = super.doGet(BASE_URL + "/players/" + playerTag)) {
                if (response.isOk()) {
                    return GsonUtil.fromJson(response.body(), ClanPlayer.class);
                } else {
                    throw new HttpException(response.body());
                }
            }
        }
    }
}

File: api/src/main/java/org/example/api/ClashApi.kt
package org.example.api

import org.example.api.pojo.Player
import cn.hutool.core.util.URLUtil
import cn.hutool.http.HttpRequest
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import org.example.api.exception.ClashApiException
import org.example.api.pojo.*
import org.example.api.tools.GsonUtil
import java.nio.charset.Charset

class ClashApi(
        private val TOKEN: String
) {

    private val moshi = Moshi
            .Builder()
            .addLast(KotlinJsonAdapterFactory())
            .build()

    companion object {
        const val CLASH_API = "https://api.clashofclans.com/v1"
    }

    /**
     * 通过部落标签获取单个部落的信息.
     * @throws ClashApiException 如果调用api发生了错误，则抛出此异常.
     */
    fun getClanInformation(clanTag: String): Clan? {
        return when (val response = this.get("$CLASH_API/clans/${encodeParam(clanTag)}")) {
            is String -> GsonUtil.fromJson(response, Clan::class.java)
            else -> null
        }
    }


    /**
     * 获取指定部落周末突袭相关信息
     * @param clanTag ClanTag
     * @throws ClashApiException 如果调用api发生了错误，则抛出此异常.
     */
    @JvmOverloads
    fun getClanCapitalRaidSeason(
            clanTag: String,
            limit: Int? = null,
            after: String? = null,
            before: String? = null
    ): ClanResult<ClanCapital> {
        val response = this.get(
                "$CLASH_API/clans/${encodeParam(clanTag)}/capitalraidseasons${
                    builderParameter(
                            limit,
                            after,
                            before
                    )
                }"
        )
        val jsonAdapter = moshi.adapter<ClanResult<ClanCapital>>(
                Types.newParameterizedType(
                        ClanResult::class.java,
                        ClanCapital::class.java
                )
        )
        return jsonAdapter.fromJson(response)!!
    }

    /**
     * 获取部落成员信息.
     * @throws ClashApiException 如果调用api发生了错误，则抛出此异常.
     */
    @JvmOverloads
    fun getClanMembers(
            clanTag: String, limit: Int? = null,
            after: String? = null,
            before: String? = null
    ): ClanResult<ClanMember> {
        val response = this.get(
                "$CLASH_API/clans/${encodeParam(clanTag)}/members${
                    builderParameter(
                            limit,
                            after,
                            before
                    )
                }"
        )
        val jsonAdapter =
                moshi.adapter<ClanResult<ClanMember>>(Types.newParameterizedType(ClanResult::class.java, ClanMember::class.java))
        return jsonAdapter.fromJson(response)!!
    }

    /**
     * 查询部落当前的联赛组信息
     * @throws ClashApiException 如果调用api发生了错误，则抛出此异常.
     */
    fun getClansCurrentWarLeaguegroup(clanTag: String): ClashWarLeagueGroup {
        val response = this.get("$CLASH_API/clans/${encodeParam(clanTag)}/currentwar/leaguegroup")
        val jsonAdapter = moshi.adapter(ClashWarLeagueGroup::class.java)
        return jsonAdapter.fromJson(response)!!
    }

    /**
     * 查询部落指定联赛场次信息
     * @throws ClashApiException 如果调用api发生了错误，则抛出此异常.
     */
    fun getClanWarLeagueInformation(warTag: String): ClanWarLeagueInfo {
        val response = this.get("$CLASH_API/clanwarleagues/wars/${encodeParam(warTag)}")
        val jsonAdapter = moshi.adapter(ClanWarLeagueInfo::class.java)
        return jsonAdapter.fromJson(response)!!
    }

    /**
     * 通过玩家标签获取玩家的信息。
     * 玩家标签可以在游戏中找到，也可以在氏族成员名单中找到。
     * 请注意，玩家标签以哈希字符'#'开头，需要正确地进行URL编码才能在URL中发挥作用。
     * 例如，玩家标签'#2ABC'在URL中会变成'%232ABC'。
     */
    fun getPlayerInformation(playerTag: String): Player? {
        val response = this.get("$CLASH_API/players/${encodeParam(playerTag)}")
        val jsonAdapter = moshi.adapter(Player::class.java)
        return jsonAdapter.fromJson(response)
    }


    private fun get(url: String): String {
        val response = HttpRequest.get(url)
                .bearerAuth(TOKEN)
                .execute()
        return when {
            response.isOk -> response.body()
            else -> throw ClashApiException(GsonUtil.fromJson(response.body(), ClashErrorResponse::class.java))
        }
    }

    private fun post(url: String): String {
        val response = HttpRequest.post(url)
                .bearerAuth(TOKEN)
                .execute()
        return when {
            response.isOk -> response.body()
            else -> throw ClashApiException(GsonUtil.fromJson(response.body(), ClashErrorResponse::class.java))
        }
    }

    /**
     * 构建查询字符串
     */
    private fun builderParameter(
            limit: Int? = null,
            after: String? = null,
            before: String? = null
    ): String {
        val params = mutableMapOf<String, Any>()
        limit?.let { i: Int -> params["limit"] = i }
        after?.let { s -> params["after"] = s }
        before?.let { s -> params["before"] = s }
        val buildQuery = URLUtil.buildQuery(params, Charset.defaultCharset())
        return if (buildQuery.isNotBlank()) "?$buildQuery" else ""
    }

    /**
     * Encoding incoming parameter
     */
    private fun encodeParam(param: String): String? {
        return URLUtil.encode(param)
    }
}